olp-cpp-sdk-core/src/http/android/HttpClient.java

```diff
@@ -329,12 +333,12 @@ protected Void doInBackground(Request... requests) {
                 status = httpConn.getResponseCode();
                 error = httpConn.getResponseMessage();
 
-                if ((status > 0)
-                    && ((status < HttpURLConnection.HTTP_OK)
-                        || (status >= HttpURLConnection.HTTP_SERVER_ERROR))) {
-                  if (retryCount++ < request.maxRetries()) {
-                    continue;
-                  }
+                if ((status == HttpURLConnection.HTTP_GATEWAY_TIMEOUT
+                        || status == HttpURLConnection.HTTP_BAD_GATEWAY
+                        || status == HttpURLConnection.HTTP_INTERNAL_ERROR
+                        || status == HttpURLConnection.HTTP_UNAVAILABLE)
+                    && retryCount++ < request.maxRetries()) {
+                  continue;
```

Comment:
What is the reason about changing the retry mechanism? I see that in `NetworkCurl` we also use the old retry mechanism, thus, your changes will make the `NetworkAndroid` retry logic inconsistent with other implementations.

Reply:
What do you mean with 'old retry mechanism'? As far as I see I didn't change the mechanism, as it will still do the "retrying". Here's what I did and my intention behind it:

- Removed the higher than zero check because it is redundant (all http codes are positive)
- Changed the status codes to replace the deprecated one (HTTP_SERVER_ERROR "should have never existed" according to android's own documentation: https://developer.android.com/reference/java/net/HttpURLConnection.html#HTTP_SERVER_ERROR) - so its replacement is HTTP_INTERNAL_ERROR.
- Finally, from my understanding when reading this code, the author's original intention was to retry the connection for failures that are potentially temporary and on the server side (as it makes no sense for example to retry if the device has no internet connection), so I made it more explicit by adding such error codes.

Though to be honest, I would vote to remove the retry mechanism altogether in a separate patch since:
- iOS implementation does not contain it.
- IMO retrying requests should not be done at this level at all.
- It could be that there is a specific case, for curl-based clients , that retrying could make sense, but then I would like to know when that is the case, if possible.

Otherwise, please let me know if I'm missing anything.

Reply:
Thanks for the detailed comment.

> What do you mean with 'old retry mechanism'?

As for the 'old retry mechanism' I meant the retry mechanism that is currently in SDK (i.e. previous version of the retry logic relative to your PR).

My concern regarding your changes is that you slightly changes current retry logic - as far as I see, you omitted 2 HTTP error codes:
- https://docs.oracle.com/javase/7/docs/api/java/net/HttpURLConnection.html#HTTP_NOT_IMPLEMENTED ;
- https://docs.oracle.com/javase/7/docs/api/java/net/HttpURLConnection.html#HTTP_VERSION ;

Even though, a client most probably will not be interested in retrying on these error codes, but still your change is not 'backward-compatible' with current retry mechanism. So, if we agree to apply your changes in HttpClient.java, we have to change retry logic in other implementations, like `NetworkCurl` (https://github.com/heremaps/here-olp-sdk-cpp/blob/master/olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp#L937). So, it would be perfect if you apply those changes in this PR as well, otherwise we will handle the `NetworkCurl` by ourselves.

> IMO retrying requests should not be done at this level at all.

We have some SPIKE about reconsider current retry mechanism, so, the retry logic might change in the future release(s).

Reply:
Okay, I see your point. So when you say backward compatibility you mean consistent behaviour across different implementations, and not backward compatibility in android (a change here (android) should have no impact whatsoever for users of the curl implementation - though, I see value in making the behaviour consistent across all implementations).

I will remove these changes, since it is not critical to the improvement I need on this patch.

Please note though that the following (current implementation):

((status > 0) && ((status < 200)|| (status >= 500)))

doesn't provide much value since:
Http status codes at 1XX range are "informational", and other error codes such as 501 (not implemented), 511, etc, cannot be recovered by simply retrying.

Reply:
Ok, we will address your suggested changes about retry logic in a separate ticket.